fix in joblib broke fscacher?

Using joblib with the recent fix (thanks @jwodder) http://github.com/joblib/joblib/commit/457d2c8a926105fd156b1dfccfaae3e500d22451 (1.0.1-3-g457d2c8) causes tests to error out with

```
E               TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
```

<details>
<summary>full log</summary> 

```shell
(git)lena:~/proj/fscacher[master]git
$> python -m pytest --tb=short src/fscacher/tests/test_cache.py
========================= test session starts ==========================
platform linux -- Python 3.9.1, pytest-6.2.2, py-1.9.0, pluggy-0.13.0
rootdir: /home/yoh/proj/fscacher, configfile: tox.ini
plugins: mock-3.5.1, cov-2.11.1, pyfakefs-4.1.0, forked-1.3.0, timeout-1.4.1, xdist-1.32.0, hypothesis-5.41.2, xonsh-0.9.24
collected 16 items                                                     

src/fscacher/tests/test_cache.py ..FFFF..........              [100%]

============================== FAILURES ==============================
_________________________ test_memoize_path __________________________
src/fscacher/tests/test_cache.py:131: in test_memoize_path
    check_new_memoread(1, "content")
src/fscacher/tests/test_cache.py:98: in check_new_memoread
    assert memoread(path, arg) == content
src/fscacher/cache.py:116: in fingerprinter
    ret = fingerprinted(path, *args, **kwargs_)
../../deb/gits/joblib/joblib/memory.py:591: in __call__
    return self._cached_call(args, kwargs)[0]
../../deb/gits/joblib/joblib/memory.py:483: in _cached_call
    func_id, args_id = self._get_output_identifiers(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:620: in _get_output_identifiers
    argument_hash = self._get_argument_hash(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:614: in _get_argument_hash
    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),
../../deb/gits/joblib/joblib/func_inspect.py:295: in filter_args
    raise TypeError("Ignore list for %s() contains an unexpected "
E   TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
_______________________ test_memoize_path_dir ________________________
src/fscacher/tests/test_cache.py:212: in test_memoize_path_dir
    check_new_memoread(1, 9)
src/fscacher/tests/test_cache.py:178: in check_new_memoread
    assert memoread(path, arg) == content
src/fscacher/cache.py:116: in fingerprinter
    ret = fingerprinted(path, *args, **kwargs_)
../../deb/gits/joblib/joblib/memory.py:591: in __call__
    return self._cached_call(args, kwargs)[0]
../../deb/gits/joblib/joblib/memory.py:483: in _cached_call
    func_id, args_id = self._get_output_identifiers(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:620: in _get_output_identifiers
    argument_hash = self._get_argument_hash(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:614: in _get_argument_hash
    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),
../../deb/gits/joblib/joblib/func_inspect.py:295: in filter_args
    raise TypeError("Ignore list for %s() contains an unexpected "
E   TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
_____________________ test_memoize_path_persist ______________________
src/fscacher/tests/test_cache.py:275: in test_memoize_path_persist
    assert outputs[0].stdout.strip().decode() == "Running script.py.DONE"
E   AssertionError: assert '' == 'Running script.py.DONE'
E     - Running script.py.DONE
------------------------ Captured stdout call ------------------------
Full outputs: [CompletedProcess(args=['/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/bin/python', '/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py'], returncode=1, stdout=b'', stderr=b'Traceback (most recent call last):\n  File "/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py", line 10, in <module>\n    print(func(r"/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py"))\n  File "/home/yoh/proj/fscacher/src/fscacher/cache.py", line 116, in fingerprinter\n    ret = fingerprinted(path, *args, **kwargs_)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 591, in __call__\n    return self._cached_call(args, kwargs)[0]\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 483, in _cached_call\n    func_id, args_id = self._get_output_identifiers(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 620, in _get_output_identifiers\n    argument_hash = self._get_argument_hash(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 614, in _get_argument_hash\n    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),\n  File "/home/yoh/deb/gits/joblib/joblib/func_inspect.py", line 295, in filter_args\n    raise TypeError("Ignore list for %s() contains an unexpected "\nTypeError: Ignore list for func() contains an unexpected keyword argument \'_cache_fingerprint\'\n'), CompletedProcess(args=['/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/bin/python', '/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py'], returncode=1, stdout=b'', stderr=b'Traceback (most recent call last):\n  File "/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py", line 10, in <module>\n    print(func(r"/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py"))\n  File "/home/yoh/proj/fscacher/src/fscacher/cache.py", line 116, in fingerprinter\n    ret = fingerprinted(path, *args, **kwargs_)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 591, in __call__\n    return self._cached_call(args, kwargs)[0]\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 483, in _cached_call\n    func_id, args_id = self._get_output_identifiers(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 620, in _get_output_identifiers\n    argument_hash = self._get_argument_hash(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 614, in _get_argument_hash\n    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),\n  File "/home/yoh/deb/gits/joblib/joblib/func_inspect.py", line 295, in filter_args\n    raise TypeError("Ignore list for %s() contains an unexpected "\nTypeError: Ignore list for func() contains an unexpected keyword argument \'_cache_fingerprint\'\n'), CompletedProcess(args=['/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/bin/python', '/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py'], returncode=1, stdout=b'', stderr=b'Traceback (most recent call last):\n  File "/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py", line 10, in <module>\n    print(func(r"/home/yoh/.tmp/pytest-of-yoh/pytest-83/test_memoize_path_persist0/script.py"))\n  File "/home/yoh/proj/fscacher/src/fscacher/cache.py", line 116, in fingerprinter\n    ret = fingerprinted(path, *args, **kwargs_)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 591, in __call__\n    return self._cached_call(args, kwargs)[0]\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 483, in _cached_call\n    func_id, args_id = self._get_output_identifiers(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 620, in _get_output_identifiers\n    argument_hash = self._get_argument_hash(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 614, in _get_argument_hash\n    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),\n  File "/home/yoh/deb/gits/joblib/joblib/func_inspect.py", line 295, in filter_args\n    raise TypeError("Ignore list for %s() contains an unexpected "\nTypeError: Ignore list for func() contains an unexpected keyword argument \'_cache_fingerprint\'\n')]
______________________ test_memoize_path_tokens ______________________
src/fscacher/tests/test_cache.py:311: in test_memoize_path_tokens
    check_new_memoread(memoread, 0, "content")
src/fscacher/tests/test_cache.py:299: in check_new_memoread
    assert call(path, arg) == content
src/fscacher/cache.py:116: in fingerprinter
    ret = fingerprinted(path, *args, **kwargs_)
../../deb/gits/joblib/joblib/memory.py:591: in __call__
    return self._cached_call(args, kwargs)[0]
../../deb/gits/joblib/joblib/memory.py:483: in _cached_call
    func_id, args_id = self._get_output_identifiers(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:620: in _get_output_identifiers
    argument_hash = self._get_argument_hash(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:614: in _get_argument_hash
    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),
../../deb/gits/joblib/joblib/func_inspect.py:295: in filter_args
    raise TypeError("Ignore list for %s() contains an unexpected "
E   TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
====================== short test summary info =======================
FAILED src/fscacher/tests/test_cache.py::test_memoize_path - TypeEr...
FAILED src/fscacher/tests/test_cache.py::test_memoize_path_dir - Ty...
FAILED src/fscacher/tests/test_cache.py::test_memoize_path_persist
(dev3) 1 66715 ->1 [2].....................................:Thu 18 Mar 2021 06:46:12 PM EDT:.
(git)lena:~/proj/fscacher[master]git
$> 
(dev3) 1 66715 ->130 [2].....................................:Thu 18 Mar 2021 06:46:13 PM EDT:.
(git)lena:~/proj/fscacher[master]git
$> python -m pytest --tb=short src/fscacher/tests/test_cache.py
====================================================== test session starts ======================================================
platform linux -- Python 3.9.1, pytest-6.2.2, py-1.9.0, pluggy-0.13.0
rootdir: /home/yoh/proj/fscacher, configfile: tox.ini
plugins: mock-3.5.1, cov-2.11.1, pyfakefs-4.1.0, forked-1.3.0, timeout-1.4.1, xdist-1.32.0, hypothesis-5.41.2, xonsh-0.9.24
collected 16 items                                                                                                              

src/fscacher/tests/test_cache.py ..FFFF..........                                                                         [100%]

=========================================================== FAILURES ============================================================
_______________________________________________________ test_memoize_path _______________________________________________________
src/fscacher/tests/test_cache.py:131: in test_memoize_path
    check_new_memoread(1, "content")
src/fscacher/tests/test_cache.py:98: in check_new_memoread
    assert memoread(path, arg) == content
src/fscacher/cache.py:116: in fingerprinter
    ret = fingerprinted(path, *args, **kwargs_)
../../deb/gits/joblib/joblib/memory.py:591: in __call__
    return self._cached_call(args, kwargs)[0]
../../deb/gits/joblib/joblib/memory.py:483: in _cached_call
    func_id, args_id = self._get_output_identifiers(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:620: in _get_output_identifiers
    argument_hash = self._get_argument_hash(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:614: in _get_argument_hash
    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),
../../deb/gits/joblib/joblib/func_inspect.py:295: in filter_args
    raise TypeError("Ignore list for %s() contains an unexpected "
E   TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
_____________________________________________________ test_memoize_path_dir _____________________________________________________
src/fscacher/tests/test_cache.py:212: in test_memoize_path_dir
    check_new_memoread(1, 9)
src/fscacher/tests/test_cache.py:178: in check_new_memoread
    assert memoread(path, arg) == content
src/fscacher/cache.py:116: in fingerprinter
    ret = fingerprinted(path, *args, **kwargs_)
../../deb/gits/joblib/joblib/memory.py:591: in __call__
    return self._cached_call(args, kwargs)[0]
../../deb/gits/joblib/joblib/memory.py:483: in _cached_call
    func_id, args_id = self._get_output_identifiers(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:620: in _get_output_identifiers
    argument_hash = self._get_argument_hash(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:614: in _get_argument_hash
    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),
../../deb/gits/joblib/joblib/func_inspect.py:295: in filter_args
    raise TypeError("Ignore list for %s() contains an unexpected "
E   TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
___________________________________________________ test_memoize_path_persist ___________________________________________________
src/fscacher/tests/test_cache.py:275: in test_memoize_path_persist
    assert outputs[0].stdout.strip().decode() == "Running script.py.DONE"
E   AssertionError: assert '' == 'Running script.py.DONE'
E     - Running script.py.DONE
----------------------------------------------------- Captured stdout call ------------------------------------------------------
Full outputs: [CompletedProcess(args=['/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/bin/python', '/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py'], returncode=1, stdout=b'', stderr=b'Traceback (most recent call last):\n  File "/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py", line 10, in <module>\n    print(func(r"/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py"))\n  File "/home/yoh/proj/fscacher/src/fscacher/cache.py", line 116, in fingerprinter\n    ret = fingerprinted(path, *args, **kwargs_)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 591, in __call__\n    return self._cached_call(args, kwargs)[0]\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 483, in _cached_call\n    func_id, args_id = self._get_output_identifiers(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 620, in _get_output_identifiers\n    argument_hash = self._get_argument_hash(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 614, in _get_argument_hash\n    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),\n  File "/home/yoh/deb/gits/joblib/joblib/func_inspect.py", line 295, in filter_args\n    raise TypeError("Ignore list for %s() contains an unexpected "\nTypeError: Ignore list for func() contains an unexpected keyword argument \'_cache_fingerprint\'\n'), CompletedProcess(args=['/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/bin/python', '/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py'], returncode=1, stdout=b'', stderr=b'Traceback (most recent call last):\n  File "/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py", line 10, in <module>\n    print(func(r"/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py"))\n  File "/home/yoh/proj/fscacher/src/fscacher/cache.py", line 116, in fingerprinter\n    ret = fingerprinted(path, *args, **kwargs_)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 591, in __call__\n    return self._cached_call(args, kwargs)[0]\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 483, in _cached_call\n    func_id, args_id = self._get_output_identifiers(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 620, in _get_output_identifiers\n    argument_hash = self._get_argument_hash(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 614, in _get_argument_hash\n    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),\n  File "/home/yoh/deb/gits/joblib/joblib/func_inspect.py", line 295, in filter_args\n    raise TypeError("Ignore list for %s() contains an unexpected "\nTypeError: Ignore list for func() contains an unexpected keyword argument \'_cache_fingerprint\'\n'), CompletedProcess(args=['/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/bin/python', '/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py'], returncode=1, stdout=b'', stderr=b'Traceback (most recent call last):\n  File "/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py", line 10, in <module>\n    print(func(r"/home/yoh/.tmp/pytest-of-yoh/pytest-84/test_memoize_path_persist0/script.py"))\n  File "/home/yoh/proj/fscacher/src/fscacher/cache.py", line 116, in fingerprinter\n    ret = fingerprinted(path, *args, **kwargs_)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 591, in __call__\n    return self._cached_call(args, kwargs)[0]\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 483, in _cached_call\n    func_id, args_id = self._get_output_identifiers(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 620, in _get_output_identifiers\n    argument_hash = self._get_argument_hash(*args, **kwargs)\n  File "/home/yoh/deb/gits/joblib/joblib/memory.py", line 614, in _get_argument_hash\n    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),\n  File "/home/yoh/deb/gits/joblib/joblib/func_inspect.py", line 295, in filter_args\n    raise TypeError("Ignore list for %s() contains an unexpected "\nTypeError: Ignore list for func() contains an unexpected keyword argument \'_cache_fingerprint\'\n')]
___________________________________________________ test_memoize_path_tokens ____________________________________________________
src/fscacher/tests/test_cache.py:311: in test_memoize_path_tokens
    check_new_memoread(memoread, 0, "content")
src/fscacher/tests/test_cache.py:299: in check_new_memoread
    assert call(path, arg) == content
src/fscacher/cache.py:116: in fingerprinter
    ret = fingerprinted(path, *args, **kwargs_)
../../deb/gits/joblib/joblib/memory.py:591: in __call__
    return self._cached_call(args, kwargs)[0]
../../deb/gits/joblib/joblib/memory.py:483: in _cached_call
    func_id, args_id = self._get_output_identifiers(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:620: in _get_output_identifiers
    argument_hash = self._get_argument_hash(*args, **kwargs)
../../deb/gits/joblib/joblib/memory.py:614: in _get_argument_hash
    return hashing.hash(filter_args(self.func, self.ignore, args, kwargs),
../../deb/gits/joblib/joblib/func_inspect.py:295: in filter_args
    raise TypeError("Ignore list for %s() contains an unexpected "
E   TypeError: Ignore list for memoread() contains an unexpected keyword argument '_cache_fingerprint'
==================================================== short test summary info ====================================================
FAILED src/fscacher/tests/test_cache.py::test_memoize_path - TypeError: Ignore list for memoread() contains an unexpected keyw...
FAILED src/fscacher/tests/test_cache.py::test_memoize_path_dir - TypeError: Ignore list for memoread() contains an unexpected ...
FAILED src/fscacher/tests/test_cache.py::test_memoize_path_persist - AssertionError: assert '' == 'Running script.py.DONE'
FAILED src/fscacher/tests/test_cache.py::test_memoize_path_tokens - TypeError: Ignore list for memoread() contains an unexpect...
================================================= 4 failed, 12 passed in 1.69s ==================================================
```
</details>

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -22,6 +22,12 @@ jobs:
                     - '3.8'
                     - '3.9'
                     - pypy3
+                toxenv:
+                    - py
+                include:
+                    - os: ubuntu-latest
+                      python-version: '3.6'
+                      toxenv: py-dev
         steps:
             - name: Check out repository
               uses: actions/checkout@v2
@@ -37,9 +43,15 @@ jobs:
                 python -m pip install --upgrade --upgrade-strategy=eager tox
 
             - name: Run tests
+              if: matrix.toxenv == 'py'
               run: tox -e py -- --cov-report=xml
 
+            - name: Run generic tests
+              if: matrix.toxenv != 'py'
+              run: tox -e ${{ matrix.toxenv }}
+
             - name: Upload coverage to Codecov
+              if: matrix.toxenv == 'py'
               uses: codecov/codecov-action@v1
               with:
                   fail_ci_if_error: false

--- a/src/fscacher/cache.py
+++ b/src/fscacher/cache.py
@@ -1,5 +1,6 @@
 from collections import deque, namedtuple
 from functools import wraps
+from inspect import Parameter, signature
 import logging
 import os
 import os.path as op
@@ -76,12 +77,23 @@ class PersistentCache(object):
         # we need to actually decorate a function
         fingerprint_kwarg = "_cache_fingerprint"
 
-        @self.memoize
         @wraps(f)  # important, so .memoize correctly caches different `f`
         def fingerprinted(path, *args, **kwargs):
             _ = kwargs.pop(fingerprint_kwarg)  # discard
             lgr.debug("Running original %s on %r", f, path)
             return f(path, *args, **kwargs)
+
+        # We need to add the fingerprint_kwarg to fingerprinted's signature so
+        # that joblib doesn't complain:
+        sig = signature(fingerprinted)
+        fp_kwarg_param = Parameter(
+            fingerprint_kwarg, Parameter.KEYWORD_ONLY, default=None
+        )
+        sig2 = sig.replace(
+            parameters=tuple(sig.parameters.values()) + (fp_kwarg_param,)
+        )
+        fingerprinted.__signature__ = sig2
+        fingerprinted = self.memoize(fingerprinted)
 
         @wraps(f)
         def fingerprinter(path, *args, **kwargs):

--- a/tox.ini
+++ b/tox.ini
@@ -10,6 +10,7 @@ deps =
     flake8-bugbear
     flake8-builtins~=1.4
     flake8-import-order-jwodder
+    dev: joblib @ git+https://github.com/joblib/joblib.git
     pytest~=6.0
     pytest-cov~=2.0
     pytest-mock~=3.0

